[Ho Lok Wah] iP

ThomasHoooo opened the pull request:
## DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!
Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ThomasHoooo committed:
Feat: Duke can echo commands

ThomasHoooo committed:
Feat: Duke can store whatever text entered by the user and display them back to the user when requested.

ThomasHoooo committed:
Feat: Duke can mark tasks as done.

ThomasHoooo committed:
Feat: Add support for tracking three types of tasks

ThomasHoooo committed:
Feat: Add support for tracking three types of tasks

ThomasHoooo committed:
Test using the I/O redirection technique

ThomasHoooo committed:
Feat: handle errors for duke

ThomasHoooo committed:
Feat: Add support for deleting tasks from the list.

ThomasHoooo committed:
Save the tasks in the hard disk automatically whenever the task list changes

ThomasHoooo committed:
Teach Duke how to understand dates and times

ThomasHoooo committed:
Merge branch 'branch-Level-7'

ThomasHoooo committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

ThomasHoooo committed:
Add a few skeleton files for OOP

ThomasHoooo committed:
Update Duke to perform more tasks

ThomasHoooo committed:
Refactor Duke to match OOP standards

ThomasHoooo committed:
Organise Duke files in packages

ThomasHoooo committed:
Merge branch 'add-gradle-support'

ThomasHoooo committed:
Add JUnit tests to test the behavior of the code.

ThomasHoooo committed:
Give users a way to find a task by searching for a keyword.

ThomasHoooo committed:
Modify Parser to meet coding standard

ThomasHoooo committed:
Add JavaDoc comments to the code.

ThomasHoooo committed:
Merge branch 'branch-A-JavaDoc'

ThomasHoooo committed:
Merge branch 'branch-A-CodingStandard'

ThomasHoooo committed:
Merge branch 'branch-Level-9'

gabyang reviewed (commented):
Wow what a great application!

ThomasHoooo committed:
Fixed bugs

ThomasHoooo committed:
Refactor regex pattern

ThomasHoooo commented:
> Wow what a great application!

Thank you for being honest!

donljh commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,36 @@
+        return (isDone ? "X" : " ");
+    }
+
+    /**
+     * Marks the task as done.
+     */
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
+    /**
+     * Unmarks the task as not done.
+     */
+    public void unmark() {
+        this.isDone = false;
+    }
```
Perhaps you could consider naming these two methods similarly since they perform similar functions? 

i.e. Renaming unmark to markAsUndone, or markAsDone to simply mark

ThomasHoooo replied:
Don't think this is possible because I have 2 commands, mark and unmark. If I combine the two methods I won't be able to check whether it's done or not done.

donljh commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,47 @@
+        tasks.get(index).unmark();
+    }
+
+    public TaskList find(String task) {
+        return new TaskList((ArrayList) tasks.stream().filter(s -> s.toString().contains(task)).collect(Collectors.<Task>toList()));
+    }
```
I see that this method tries to find a keyword that each task might contain in its description. Perhaps it might be clearer to name the relevant method parameter that way?

i.e. Rename task to keyword in the method parameter

donljh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,82 @@
+                storage.writeToFile(tasks);
+                continue;
+            case UNMARK:
+                tasks.unmark(Integer.parseInt(command.replace("mark ", "").trim()));
```
Not sure if this was a mistake, but was your intention to replace the "unmark" word instead?

ThomasHoooo replied:
Good point

donljh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,82 @@
+            switch (Parser.parse(command)) {
+            case TODO: {
+                Todo todo = new Todo(command.replace("todo", "").trim(), false);
+                tasks.add(todo);
+                Ui.printAddedMessage(todo, tasks.getSize());
+                storage.writeToFile(tasks);
+                continue;
+            }
+            case DEADLINE: {
+                String[] splitStr = command.trim().split("/by");
+                String date = splitStr[1].replace("by", "").trim();
+                Deadline deadline = new Deadline(splitStr[0].replace("deadline", "").trim(),
+                        false, date);
+                tasks.add(deadline);
+                Ui.printAddedMessage(deadline, tasks.getSize());
+                storage.writeToFile(tasks);
+                continue;
+            }
+            case EVENT: {
+                String[] splitStr = command.trim().split("/at");
+                String date = splitStr[1].replace("at", "").trim();
+                Event event = new Event(splitStr[0].replace("event", "").trim(), false, date);
+                tasks.add(event);
+                Ui.printAddedMessage(event, tasks.getSize());
+                storage.writeToFile(tasks);
+                continue;
+            }
+            case DELETE: {
+                String[] splitStr = command.trim().split("\\s+");
+                int deleteItem = Integer.parseInt(splitStr[splitStr.length - 1]) - 1;
+                Task deletedTask = tasks.remove(deleteItem);
+                Ui.printDeletedMessage(deletedTask, tasks.getSize());
+                storage.writeToFile(tasks);
+                continue;
+            }
+            case MARK:
+                tasks.markDone(Integer.parseInt(command.replace("mark ", "").trim()));
+                storage.writeToFile(tasks);
+                continue;
+            case UNMARK:
```
I like that you used the switch clause to effectively separate the logic of each command. However, some commands seem to require parsing the user input in similar fashions (mark and unmark, deadline and event). 

Perhaps it might be better to abstract the parsing, as well as each command into their own classes to better utilize OOP?

ThomasHoooo committed:
Update build.gradle to include javaFx

ThomasHoooo committed:
Add a GUI to Duke

ThomasHoooo committed:
Merge branch 'branch-Level-10'

ThomasHoooo committed:
Use assert feature to document important assumptions that should hold at various points in the code.

ThomasHoooo committed:
Merge pull request #1 from ThomasHoooo/branch-A-Assertions

Use assert feature to document important assumptions that should hold…

ThomasHoooo committed:
Remove meaningless comments

ThomasHoooo committed:
Merge pull request #2 from ThomasHoooo/branch-A-CodeQuality

Remove meaningless comments

ThomasHoooo committed:
Add update function to update toDoList

ThomasHoooo committed:
Merge pull request #3 from ThomasHoooo/branch-C-Update

Add update function to update toDoList

ThomasHoooo committed:
Fix minor bugs

ThomasHoooo committed:
Fix: write updated tasks into hard disk storage

Updated tasks were not written into hard disk storage.

An updated storage in the hard disk is needed after user updated his or her tasks.

Let's,
* fix the bug to allow user update tasks in the hard disk

ThomasHoooo committed:
Fix path of application in build.gradle

ThomasHoooo committed:
Add initialise message

ThomasHoooo committed:
Update README.md

ThomasHoooo committed:
Close the application after saying bye

ThomasHoooo committed:
Remove all unused functions

ThomasHoooo committed:
Add application screenshot
